### PR TITLE
Update link dropdown menus to use new Tippy dropdowns

### PR DESCRIPTION
**Pull request by janfaracik**
This PR replaces the YUI menus used for link dropdowns with a modern framework, Tippy (introduced in #6408). The menus feel more modern and fit in better with the ongoing Jenkins redesign, as well as cut down on our reliance on YUI.

**Some screenshots**

image

image

image

**What this isn't**
Not all areas have been updated to use the dropdowns, such as the button dropdowns commonly found in Manage Jenkins/Config pages. This will be done in a followup PR to keep this one simpler. 

I'd also like to introduce a simple API for plugin developers to create their own dropdowns with ease, this again will be looked at in a future PR.

### Testing done

On a a page with link dropdown menus (such as Dashboard), click the chevron to open the menu. Notice the new menu pops up in a similar fashion to before. 

On open menus use keyboard up and down arrows to navigate through the items, pressing enter to activate an item.

Interact with dropdown items that aren't links, and are instead actions, such as the Delete item which opens a modal.

Go to the Configure System page and interact with the page's custom breadcrumb, all the sections previously displayed should be there again with a new style.

### Proposed changelog entries

- Update appearance and framework for link dropdown menus

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7474"></a>

### CODE REVIEW

**Comment by daniel-beck** [comment]
What happens when you make the window smaller than the dropdown (e.g. Manage Jenkins) would be high, i.e. when the existing dropdowns' scrolling feature is needed?

**Comment by janfaracik** [comment]
> What happens when you make the window smaller than the dropdown (e.g. Manage Jenkins) would be high, i.e. when the existing dropdowns' scrolling feature is needed?

I hope I'm understanding you correctly, if the browser is too small the dropdown will display a scrollbar:

https://user-images.githubusercontent.com/43062514/206275173-d3841888-071a-4eb8-a6e2-eb7497bdcf21.mov

**Comment by daniel-beck** [comment]
> > What happens when you make the window smaller than the dropdown (e.g. Manage Jenkins) would be high, i.e. when the existing dropdowns' scrolling feature is needed?
> 
> I hope I'm understanding you correctly, if the browser is too small the dropdown will display a scrollbar:

Thanks, that's what I was asking about. Looks like a great solution, much better than the weird "buttons" we currently have (FWIW I would have run this rather than commented, but it didn't build at the time.)

**Review by timja — COMMENTED** [review]
I tested in Chrome a number of different pages and dropdowns. I had a fairly simple setup as just using the core-pr-tester, I found one issue for accessibility, happy with the rest of it

----

ATH is passing

----

Keyboard navigation to the right doesn't seem to work, e.g. from the dashboard dropdown menu:

image

This does work correctly on the latest weekly

**Review by timja — APPROVED** [review]
LGTM

**Review by NotMyFault — APPROVED** [review]

**Review by yaroslavafenkin — APPROVED** [review]
Seems fine to me from security perspective. Clicked around for a bit, couldn't break it.

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by daniel-beck** [comment]
Previously, empty menus weren't rendered, now they are:

> Screenshot 2023-01-05 at 14 12 34

(Arguable there shouldn't be a menu indicator with no menu contents, but that would require requesting all menus beforehand.)

----

The newly added closing animation makes in-page navigation from breadcrumb entries look weird, as the menu moves while closing due to the header disappearing.

(No recording, Quicktime Player is broken for some reason.)

----

Could you explain the rationale for translucency? This seems just distracting for no clear benefit:

> Screenshot 2023-01-05 at 14 14 12

----

FTR opening/closing context menus appears to use less CPU than before, so that's a nice improvement for lower powered devices!

**Review by daniel-beck — CHANGES_REQUESTED** [review]
There's some double-escaping going on:

> Screenshot 2023-01-05 at 14 30 50

Before:

> Screenshot 2023-01-05 at 14 32 02

**Comment by janfaracik** [comment]
> > Screenshot 2023-01-05 at 14 30 50

Removed Util.escape(..) when setting the display name for a menu item, the escaping is done just on the frontend now. Is that all good with you?

image

**Comment by janfaracik** [comment]
> I have not yet confirmed the output isn't used elsewhere where the API response would expect to be provided pre-escaped. While unlikely, it'd be annoying to introduce a vulnerability this way elsewhere.

Hey, not to push - any update on this?

**Comment by daniel-beck** [comment]
I don't expect to have time for at least several more days. Others are welcome to look into this though.

**Comment by janfaracik** [comment]
> Perhaps a menu could be opened with a small non-interactive entry saying it's empty? But the weird circle looks broken.

Added - thanks!

> The newly added closing animation makes in-page navigation from breadcrumb entries look weird, as the menu moves while closing due to the header disappearing.

I'm not sure I understand/can't reproduce on my end.

> Could you explain the rationale for translucency? This seems just distracting for no clear benefit:

Just a design decision, it does look a tad odd there due to the extreme contrast between the header and background, but still legible.

**Comment by uhafner** [comment]
Is it possible that the chevron stays opened while the menu is shown? It is already somewhat disturbing that the breadcrumb text moves horizontally while hovering, this is now even more problematic. When the menu opens the chevron disappears.

**Comment by janfaracik** [comment]
> Would it be possible that the chevron stays open while the menu is shown? It is already somewhat disturbing that the breadcrumb text moves horizontally while hovering, this is now even more problematic. When the menu opens the chevron disappears.

That's a good idea and I've just implemented it - 

image

Would be good to revisit the breadcrumbs dropdowns in the future as I also agree the movement is problematic.

**Review by uhafner — APPROVED** [review]
This looks better now, when the breadcrumbs stay selected. Thanks for fixing!

File: war/src/main/js/components/dropdowns/jumplists.js
Comment (janfaracik) [inline war/src/main/js/components/dropdowns/jumplists.js:103]: This needs to be updated to reflect #7727

**Comment by timja** [comment]
Lots of test failures

**Comment by janfaracik** [comment]
> I don't expect to have time for at least several more days. Others are welcome to look into this though.

Doing a cursory search on GitHub yields 1k+ results for usages of `displayName` (not necessarily relating to `MenuItem`) 

https://github.com/search?q=org%3Ajenkinsci+displayName+NOT+repo%3Ajenkinsci%2Fjenkins+language%3Axml&type=code&l=Groovy

https://github.com/search?q=org%3Ajenkinsci+displayName+NOT+repo%3Ajenkinsci%2Fjenkins+language%3Axml&type=code

Is there a better query to search with to find potential usages in plugins? AFAIK there aren't any usages in Jenkins for it outside of the dropdown menu code.

**Comment by daniel-beck** [comment]
> Is there a better query to search with to find potential usages in plugins? AFAIK there aren't any usages in Jenkins for it outside of the dropdown menu code.

This is more specifically about callers for the context menu. In my own search I was unable to find any.

The issues I mentioned in reviews were addressed. I like the new "no items" one in particular, thanks a lot for that!

**Review by daniel-beck — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by timja** [comment]
@janfaracik you able to take a look at this conflict please?

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by daniel-beck** [comment]
Likely cause of [JENKINS-71846](https://issues.jenkins.io/browse/JENKINS-71846).

**Comment by daniel-beck** [comment]
Caused [JENKINS-72745](https://issues.jenkins.io/browse/JENKINS-72745)

**Pull request merged**
